jsonschema-validation-output-machines.md

```diff
@@ -0,0 +1,587 @@
+
+Implementations MAY elect to provide additional information.
+
+### Validation Result {#validation-result}
+
+The validation result is a boolean that indicates whether the local instance passed validation by the local subschema.
+
+The JSON key for these additional results is `valid`.
+
+### Results from Subschemas
+
+Evaluation results generated by applying a subschema to the instance or a child of the instance. Keywords which have multiple subschemas (e.g. `anyOf`) will generally generate an output unit for each subschema.  In order to accommodate potentially multiple results, the value of this property MUST be an array of output units, even if only a single output unit is produced.
+
+For "list", this property will appear only at the root output unit and will hold all output units in a flat list.
+
+For "hierarchical", this property will contain results in a tree structure where each output unit may itself have further nested results.
+
+The sequence of output units within this list is not specified and MAY be determined by the implementation.  Sets of output units are considered equivalent if they contain the same units, in any order.
+
+<!-- Not sure if it's necessary to mention equivalence. -->
+
+The JSON key for these additional results is `details`.
+
+## Output Structure {#output-structure}
+
+This specification defines three output formats.
+
+- **Flag** - A boolean which simply indicates the overall validation result with no further details.
+- **List** - Provides validation information in a flat list structure.
+- **Hierarchical** - Provides validation information in a hierarchical structure that follows the evaluation paths generated while processing the schema.
+
+An implementation MUST provide the "flag" format and SHOULD provide at least one of the "list" or "hierarchical" formats. Implementations SHOULD specify in their documentation which formats they support.
+
+For these examples, the following schema and instances will be used.
+
+```jsonc
+// schema
+{
+  "$schema": "https://json-schema.org/draft/next/schema",
+  "$id": "https://json-schema.org/schemas/example",
+  "type": "object",
+  "title": "root",
+  "properties": {
+    "foo": {
+      "allOf": [
+        { "required": [ "unspecified-prop" ] },
+        {
+          "type": "object",
+          "title": "foo-title",
+          "properties": {
+            "foo-prop": {
+              "const": 1,
+              "title": "foo-prop-title"
+            }
+          },
+          "additionalProperties": { "type": "boolean" }
+        }
+      ]
+    },
+    "bar": {
+      "$ref": "#/$defs/bar"
+    }
+  },
+  "$defs": {
+    "bar": {
+      "type": "object",
+      "title": "bar-title",
+      "properties": {
+        "bar-prop": {
+          "type": "integer",
+          "minimum": 10,
+          "title": "bar-prop-title"
+        }
+      }
+    }
+  }
+}
+
+// failing instance
+{
+  "foo": {"foo-prop": "not 1", "other-prop": false},
+  "bar": {"bar-prop": 2}
```

Comment:
Spaces for `{}` but not for `[]`?  Seems inconsistent.

Reply:
I agree that it's inconsistent, but my impressing is that that's the most common convention by a wide margin. I know that's just my experience and could be biased, so I asked ChatGPT to see what it would do. It used exactly the style I was expecting, two space indent, padded objects, and not padded arrays.

![Screenshot from 2023-08-30 16-34-12](https://github.com/json-schema-org/json-schema-spec/assets/716571/8039f832-83b8-42dd-9131-d8b87ea8b16a)